File: python/hyperliquid_sdk/hypercore.py
"""
HyperCore JSON-RPC API Client — Blocks, trading, and real-time data.

Access HyperCore-specific methods via QuickNode's /hypercore endpoint.

Example:
    >>> from hyperliquid_sdk import HyperCore
    >>> hc = HyperCore("https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN")
    >>> print(hc.latest_block_number())  # Get latest block
    >>> print(hc.latest_trades(count=10))  # Get recent trades
"""

from __future__ import annotations
from typing import Optional, List, Dict, Any
from urllib.parse import urlparse

import requests

from .errors import HyperliquidError


class HyperCore:
    """
    HyperCore JSON-RPC API — Blocks, trading, and real-time data.

    Access block data, trading operations, and real-time streams via JSON-RPC.

    Examples:
        hc = HyperCore("https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN")

        # Block data
        hc.latest_block_number()
        hc.get_block(12345)
        hc.get_batch_blocks([100, 101, 102])
        hc.latest_blocks(count=10)

        # Recent data (alternative to Info methods not on QuickNode)
        hc.latest_trades(count=10)
        hc.latest_orders(count=10)
        hc.latest_book_updates(count=10)

        # Discovery
        hc.list_dexes()
        hc.list_markets()
        hc.list_markets(dex="hyperliquidity")
    """

    def __init__(self, endpoint: str, *, timeout: int = 30):
        """
        Initialize the HyperCore client.

        Args:
            endpoint: Hyperliquid endpoint URL (e.g., https://your-endpoint.quiknode.pro/TOKEN)
            timeout: Request timeout in seconds (default: 30)
        """
        self._hypercore_url = self._build_url(endpoint)
        self._timeout = timeout
        self._session = requests.Session()
        self._id = 0

    def _build_url(self, url: str) -> str:
        """Build the /hypercore endpoint URL."""
        parsed = urlparse(url)
        base = f"{parsed.scheme}://{parsed.netloc}"
        path_parts = [p for p in parsed.path.strip("/").split("/") if p]

        # Find the token (not a known path)
        token = None
        for part in path_parts:
            if part not in ("info", "hypercore", "evm", "nanoreth", "ws"):
                token = part
                break

        if token:
            return f"{base}/{token}/hypercore"
        return f"{base}/hypercore"

    def _rpc(self, method: str, params: Any = None) -> Any:
        """Make a JSON-RPC 2.0 request."""
        self._id += 1
        body: Dict[str, Any] = {
            "jsonrpc": "2.0",
            "method": method,
            "id": self._id,
        }
        if params is not None:
            body["params"] = params
        else:
            body["params"] = {}

        try:
            resp = self._session.post(self._hypercore_url, json=body, timeout=self._timeout)
        except requests.exceptions.Timeout:
            raise HyperliquidError(
                f"Request timed out after {self._timeout}s",
                code="TIMEOUT",
                raw={"method": method, "timeout": self._timeout},
            )
        except requests.exceptions.ConnectionError as e:
            raise HyperliquidError(
                f"Connection failed: {e}",
                code="CONNECTION_ERROR",
                raw={"method": method, "error": str(e)},
            ) from e

        if resp.status_code != 200:
            raise HyperliquidError(
                f"Request failed with status {resp.status_code}",
                code="HTTP_ERROR",
                raw={"status": resp.status_code, "body": resp.text},
            )

        try:
            data = resp.json()
        except ValueError:
            raise HyperliquidError(
                "Invalid JSON response",
                code="PARSE_ERROR",
                raw={"body": resp.text[:500]},
            )

        if "error" in data:
            error = data["error"]
            # Handle both dict and string error formats
            if isinstance(error, dict):
                message = error.get("message", str(error))
                code = error.get("code", "RPC_ERROR")
            else:
                message = str(error)
                code = "RPC_ERROR"
            raise HyperliquidError(message, code=str(code), raw=error)

        return data.get("result")

    # ═══════════════════════════════════════════════════════════════════════════
    # BLOCK DATA
    # ═══════════════════════════════════════════════════════════════════════════

    def latest_block_number(self, stream: str = "trades") -> int:
        """
        Get the latest block number for a stream.

        Args:
            stream: Stream type ("trades", "orders", "events", "book", "twap", "writer_actions")
        """
        return self._rpc("hl_getLatestBlockNumber", {"stream": stream})

    def get_block(self, block_number: int, stream: str = "trades") -> Dict[str, Any]:
        """
        Get a specific block by number.

        Args:
            block_number: Block number to fetch
            stream: Stream type ("trades", "orders", "events", "book", "twap", "writer_actions")
        """
        # Uses array format: [stream, block_number]
        return self._rpc("hl_getBlock", [stream, block_number])

    def get_batch_blocks(self, from_block: int, to_block: int, stream: str = "trades") -> List[Dict[str, Any]]:
        """
        Get a range of blocks.

        Args:
            from_block: Starting block number
            to_block: Ending block number (inclusive)
            stream: Stream type ("trades", "orders", "events", "book", "twap", "writer_actions")
        """
        return self._rpc("hl_getBatchBlocks", {"stream": stream, "from": from_block, "to": to_block})

    def latest_blocks(self, stream: str = "trades", *, count: int = 10) -> Dict[str, Any]:
        """
        Get the latest blocks for a stream.

        Args:
            stream: Stream type ("trades", "orders", "book_updates", "twap", "events", "writer_actions")
            count: Number of blocks to return (default: 10)
        """
        return self._rpc("hl_getLatestBlocks", {"stream": stream, "count": count})

    # ═══════════════════════════════════════════════════════════════════════════
    # RECENT DATA (Alternative to unsupported Info methods)
    # ═══════════════════════════════════════════════════════════════════════════

    def latest_trades(self, *, count: int = 10, coin: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Get recent trades from latest blocks.

        This is an alternative to Info.recent_trades() which is not available on QuickNode.

        Args:
            count: Number of blocks to fetch (default: 10)
            coin: Optional coin filter (e.g., "BTC", "ETH")

        Returns:
            List of trade events from recent blocks
        """
        result = self.latest_blocks("trades", count=count)
        trades = []
        for block in result.get("blocks", []):
            for event in block.get("events", []):
                if len(event) >= 2:
                    user, trade = event[0], event[1]
                    if coin is None or trade.get("coin") == coin:
                        trades.append({"user": user, **trade})
        return trades

    def latest_orders(self, *, count: int = 10, coin: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Get recent order events from latest blocks.

        Args:
            count: Number of blocks to fetch (default: 10)
            coin: Optional coin filter

        Returns:
            List of order events from recent blocks
        """
        result = self.latest_blocks("orders", count=count)
        orders = []
        for block in result.get("blocks", []):
            for event in block.get("events", []):
                if len(event) >= 2:
                    user, order = event[0], event[1]
                    if coin is None or order.get("coin") == coin:
                        orders.append({"user": user, **order})
        return orders

    def latest_book_updates(self, *, count: int = 10, coin: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Get recent book updates from latest blocks.

        This is an alternative to Info.l2_book() for real-time updates.

        Args:
            count: Number of blocks to fetch (default: 10)
            coin: Optional coin filter

        Returns:
            List of book update events from recent blocks
        """
        result = self.latest_blocks("book_updates", count=count)
        updates = []
        for block in result.get("blocks", []):
            for event in block.get("events", []):
                if isinstance(event, dict):
                    if coin is None or event.get("coin") == coin:
                        updates.append(event)
        return updates

    # ═══════════════════════════════════════════════════════════════════════════
    # DISCOVERY
    # ═══════════════════════════════════════════════════════════════════════════

    def list_dexes(self) -> List[Dict[str, Any]]:
        """List all available DEXes."""
        return self._rpc("hl_listDexes")

    def list_markets(self, *, dex: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        List available markets.

        Args:
            dex: Optional DEX filter (e.g., "hyperliquidity")
        """
        params = {}
        if dex:
            params["dex"] = dex
        return self._rpc("hl_listMarkets", params if params else None)

    # ═══════════════════════════════════════════════════════════════════════════
    # ORDER QUERIES
    # ═══════════════════════════════════════════════════════════════════════════

    def open_orders(self, user: str) -> List[Dict[str, Any]]:
        """Get open orders for a user."""
        return self._rpc("hl_openOrders", {"user": user})

    def order_status(self, user: str, oid: int) -> Dict[str, Any]:
        """Get status of a specific order."""
        return self._rpc("hl_orderStatus", {"user": user, "oid": oid})

    def preflight(
        self,
        coin: str,
        is_buy: bool,
        limit_px: str,
        sz: str,
        user: str,
        *,
        reduce_only: bool = False,
        order_type: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Validate an order before signing (preflight check).

        Args:
            coin: Trading pair (e.g., "BTC")
            is_buy: True for buy, False for sell
            limit_px: Limit price as string
            sz: Size as string
            user: User address
            reduce_only: Whether order is reduce-only
            order_type: Optional order type configuration
        """
        params = {
            "coin": coin,
            "isBuy": is_buy,
            "limitPx": limit_px,
            "sz": sz,
            "user": user,
            "reduceOnly": reduce_only,
        }
        if order_type:
            params["orderType"] = order_type
        return self._rpc("hl_preflight", params)

    # ═══════════════════════════════════════════════════════════════════════════
    # BUILDER FEE
    # ═══════════════════════════════════════════════════════════════════════════

    def get_max_builder_fee(self, user: str, builder: str) -> Dict[str, Any]:
        """Get maximum builder fee for a user-builder pair."""
        return self._rpc("hl_getMaxBuilderFee", {"user": user, "builder": builder})

    # ═══════════════════════════════════════════════════════════════════════════
    # ORDER BUILDING (Returns unsigned actions for signing)
    # ═══════════════════════════════════════════════════════════════════════════

    def build_order(
        self,
        coin: str,
        is_buy: bool,
        limit_px: str,
        sz: str,
        user: str,
        *,
        reduce_only: bool = False,
        order_type: Optional[Dict[str, Any]] = None,
        cloid: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Build an order for signing.

        Args:
            coin: Trading pair (e.g., "BTC")
            is_buy: True for buy, False for sell
            limit_px: Limit price as string
            sz: Size as string
            user: User address
            reduce_only: Whether order is reduce-only
            order_type: Optional order type configuration
            cloid: Optional client order ID
        """
        params: Dict[str, Any] = {
            "coin": coin,
            "isBuy": is_buy,
            "limitPx": limit_px,
            "sz": sz,
            "user": user,
            "reduceOnly": reduce_only,
        }
        if order_type:
            params["orderType"] = order_type
        if cloid:
            params["cloid"] = cloid
        return self._rpc("hl_buildOrder", params)

    def build_cancel(self, coin: str, oid: int, user: str) -> Dict[str, Any]:
        """
        Build a cancel action for signing.

        Args:
            coin: Trading pair
            oid: Order ID to cancel
            user: User address
        """
        return self._rpc("hl_buildCancel", {"coin": coin, "oid": oid, "user": user})

    def build_modify(
        self,
        coin: str,
        oid: int,
        user: str,
        *,
        limit_px: Optional[str] = None,
        sz: Optional[str] = None,
        is_buy: Optional[bool] = None,
    ) -> Dict[str, Any]:
        """
        Build a modify action for signing.

        Args:
            coin: Trading pair
            oid: Order ID to modify
            user: User address
            limit_px: New limit price
            sz: New size
            is_buy: New side
        """
        params: Dict[str, Any] = {"coin": coin, "oid": oid, "user": user}
        if limit_px is not None:
            params["limitPx"] = limit_px
        if sz is not None:
            params["sz"] = sz
        if is_buy is not None:
            params["isBuy"] = is_buy
        return self._rpc("hl_buildModify", params)

    def build_approve_builder_fee(
        self, user: str, builder: str, max_fee_rate: str, nonce: int
    ) -> Dict[str, Any]:
        """
        Build a builder fee approval for signing.

        Args:
            user: User address
            builder: Builder address
            max_fee_rate: Maximum fee rate (e.g., "0.001" for 0.1%)
            nonce: Nonce for the action
        """
        return self._rpc("hl_buildApproveBuilderFee", {
            "user": user,
            "builder": builder,
            "maxFeeRate": max_fee_rate,
            "nonce": nonce,
        })

    def build_revoke_builder_fee(self, user: str, builder: str, nonce: int) -> Dict[str, Any]:
        """
        Build a builder fee revocation for signing.

        Args:
            user: User address
            builder: Builder address to revoke
            nonce: Nonce for the action
        """
        return self._rpc("hl_buildRevokeBuilderFee", {
            "user": user,
            "builder": builder,
            "nonce": nonce,
        })

    # ═══════════════════════════════════════════════════════════════════════════
    # SENDING SIGNED ACTIONS
    # ═══════════════════════════════════════════════════════════════════════════

    def send_order(self, action: Dict[str, Any], signature: str, nonce: int) -> Dict[str, Any]:
        """
        Send a signed order.

        Args:
            action: The order action from build_order
            signature: EIP-712 signature
            nonce: Nonce used in signing
        """
        return self._rpc("hl_sendOrder", {"action": action, "signature": signature, "nonce": nonce})

    def send_cancel(self, action: Dict[str, Any], signature: str, nonce: int) -> Dict[str, Any]:
        """
        Send a signed cancel.

        Args:
            action: The cancel action from build_cancel
            signature: EIP-712 signature
            nonce: Nonce used in signing
        """
        return self._rpc("hl_sendCancel", {"action": action, "signature": signature, "nonce": nonce})

    def send_modify(self, action: Dict[str, Any], signature: str, nonce: int) -> Dict[str, Any]:
        """
        Send a signed modify.

        Args:
            action: The modify action from build_modify
            signature: EIP-712 signature
            nonce: Nonce used in signing
        """
        return self._rpc("hl_sendModify", {"action": action, "signature": signature, "nonce": nonce})

    def send_approval(self, action: Dict[str, Any], signature: str) -> Dict[str, Any]:
        """
        Send a signed builder fee approval.

        Args:
            action: The approval action from build_approve_builder_fee
            signature: EIP-712 signature
        """
        return self._rpc("hl_sendApproval", {"action": action, "signature": signature})

    def send_revocation(self, action: Dict[str, Any], signature: str) -> Dict[str, Any]:
        """
        Send a signed builder fee revocation.

        Args:
            action: The revocation action from build_revoke_builder_fee
            signature: EIP-712 signature
        """
        return self._rpc("hl_sendRevocation", {"action": action, "signature": signature})

    # ═══════════════════════════════════════════════════════════════════════════
    # WEBSOCKET SUBSCRIPTIONS (via JSON-RPC)
    # ═══════════════════════════════════════════════════════════════════════════

    def subscribe(self, subscription: Dict[str, Any]) -> Dict[str, Any]:
        """
        Subscribe to a WebSocket stream via JSON-RPC.

        Args:
            subscription: Subscription parameters (type, coin, user, etc.)

        Example:
            hc.subscribe({"type": "trades", "coin": "BTC"})
        """
        return self._rpc("hl_subscribe", {"subscription": subscription})

    def unsubscribe(self, subscription: Dict[str, Any]) -> Dict[str, Any]:
        """
        Unsubscribe from a WebSocket stream.

        Args:
            subscription: Subscription parameters to unsubscribe from
        """
        return self._rpc("hl_unsubscribe", {"subscription": subscription})

    def __enter__(self) -> HyperCore:
        return self

    def __exit__(self, *args) -> None:
        self._session.close()

    def __repr__(self) -> str:
        return f"<HyperCore {self._hypercore_url[:40]}...>"

File: go/hyperliquid/hypercore.go
package hyperliquid

import (
	"context"
)

// HyperCoreClient provides access to the HyperCore JSON-RPC API.
type HyperCoreClient struct {
	url  string
	http *HTTPClient
}

// NewHyperCoreClient creates a new HyperCore API client.
func NewHyperCoreClient(endpoint string, http *HTTPClient) *HyperCoreClient {
	return &HyperCoreClient{
		url:  buildHyperCoreURL(endpoint),
		http: http,
	}
}

func (c *HyperCoreClient) call(ctx context.Context, method string, params any) (any, error) {
	body := map[string]any{
		"jsonrpc": "2.0",
		"method":  method,
		"params":  params,
		"id":      1,
	}

	result, err := c.http.PostRaw(ctx, c.url, body)
	if err != nil {
		return nil, err
	}

	if m, ok := result.(map[string]any); ok {
		if errVal, ok := m["error"]; ok && errVal != nil {
			if errMap, ok := errVal.(map[string]any); ok {
				msg := ""
				if m, ok := errMap["message"].(string); ok {
					msg = m
				}
				return nil, NewError(ErrorCodeHTTPError, msg).WithRaw(errMap)
			}
		}
		if res, ok := m["result"]; ok {
			return res, nil
		}
	}

	return result, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// BLOCK QUERIES
// ═══════════════════════════════════════════════════════════════════════════════

// LatestBlockNumber returns the latest block number for a stream.
// Stream options: "trades", "orders", "events", "book", "twap", "writer_actions"
func (c *HyperCoreClient) LatestBlockNumber(stream ...string) (int64, error) {
	s := "trades"
	if len(stream) > 0 && stream[0] != "" {
		s = stream[0]
	}
	result, err := c.call(context.Background(), "hl_getLatestBlockNumber", map[string]any{"stream": s})
	if err != nil {
		return 0, err
	}
	if n, ok := result.(float64); ok {
		return int64(n), nil
	}
	return 0, nil
}

// GetBlock returns block data by number.
// Stream options: "trades", "orders", "events", "book", "twap", "writer_actions"
func (c *HyperCoreClient) GetBlock(blockNumber int64, stream ...string) (map[string]any, error) {
	s := "trades"
	if len(stream) > 0 && stream[0] != "" {
		s = stream[0]
	}
	// Uses array format: [stream, block_number]
	result, err := c.call(context.Background(), "hl_getBlock", []any{s, blockNumber})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// GetBatchBlocks returns a range of blocks.
// Stream options: "trades", "orders", "events", "book", "twap", "writer_actions"
func (c *HyperCoreClient) GetBatchBlocks(fromBlock, toBlock int64, stream ...string) ([]any, error) {
	s := "trades"
	if len(stream) > 0 && stream[0] != "" {
		s = stream[0]
	}
	result, err := c.call(context.Background(), "hl_getBatchBlocks", map[string]any{
		"stream": s,
		"from":   fromBlock,
		"to":     toBlock,
	})
	if err != nil {
		return nil, err
	}
	if arr, ok := result.([]any); ok {
		return arr, nil
	}
	return nil, nil
}

// LatestBlocks returns the latest N blocks for a stream.
// Stream options: "trades", "orders", "book_updates", "twap", "events", "writer_actions"
func (c *HyperCoreClient) LatestBlocks(stream string, count int) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_getLatestBlocks", map[string]any{
		"stream": stream,
		"count":  count,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// RECENT DATA (Alternative to unsupported Info methods)
// ═══════════════════════════════════════════════════════════════════════════════

// LatestTrades returns recent trades from latest blocks.
// This is an alternative to Info.RecentTrades() which may not be available on QuickNode.
func (c *HyperCoreClient) LatestTrades(count int, coin string) ([]map[string]any, error) {
	result, err := c.LatestBlocks("trades", count)
	if err != nil {
		return nil, err
	}

	var trades []map[string]any
	blocks, _ := result["blocks"].([]any)
	for _, block := range blocks {
		blockMap, ok := block.(map[string]any)
		if !ok {
			continue
		}
		events, _ := blockMap["events"].([]any)
		for _, event := range events {
			eventArr, ok := event.([]any)
			if !ok || len(eventArr) < 2 {
				continue
			}
			user, _ := eventArr[0].(string)
			trade, ok := eventArr[1].(map[string]any)
			if !ok {
				continue
			}
			// Filter by coin if specified
			if coin != "" {
				tradeCoin, _ := trade["coin"].(string)
				if tradeCoin != coin {
					continue
				}
			}
			// Add user to trade data
			tradeCopy := make(map[string]any)
			tradeCopy["user"] = user
			for k, v := range trade {
				tradeCopy[k] = v
			}
			trades = append(trades, tradeCopy)
		}
	}
	return trades, nil
}

// LatestOrders returns recent order events from latest blocks.
func (c *HyperCoreClient) LatestOrders(count int, coin string) ([]map[string]any, error) {
	result, err := c.LatestBlocks("orders", count)
	if err != nil {
		return nil, err
	}

	var orders []map[string]any
	blocks, _ := result["blocks"].([]any)
	for _, block := range blocks {
		blockMap, ok := block.(map[string]any)
		if !ok {
			continue
		}
		events, _ := blockMap["events"].([]any)
		for _, event := range events {
			// Orders stream returns dict events with 'user', 'order', 'status' fields
			eventMap, ok := event.(map[string]any)
			if !ok {
				continue
			}
			orderData, ok := eventMap["order"].(map[string]any)
			if !ok {
				continue
			}
			// Filter by coin if specified
			if coin != "" {
				orderCoin, _ := orderData["coin"].(string)
				if orderCoin != coin {
					continue
				}
			}
			// Build order entry with user and status
			orderCopy := make(map[string]any)
			orderCopy["user"] = eventMap["user"]
			orderCopy["status"] = eventMap["status"]
			for k, v := range orderData {
				orderCopy[k] = v
			}
			orders = append(orders, orderCopy)
		}
	}
	return orders, nil
}

// LatestBookUpdates returns recent book updates from latest blocks.
func (c *HyperCoreClient) LatestBookUpdates(count int, coin string) ([]map[string]any, error) {
	result, err := c.LatestBlocks("book_updates", count)
	if err != nil {
		return nil, err
	}

	var updates []map[string]any
	blocks, _ := result["blocks"].([]any)
	for _, block := range blocks {
		blockMap, ok := block.(map[string]any)
		if !ok {
			continue
		}
		events, _ := blockMap["events"].([]any)
		for _, event := range events {
			eventMap, ok := event.(map[string]any)
			if !ok {
				continue
			}
			// Filter by coin if specified
			if coin != "" {
				eventCoin, _ := eventMap["coin"].(string)
				if eventCoin != coin {
					continue
				}
			}
			updates = append(updates, eventMap)
		}
	}
	return updates, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// MARKET DISCOVERY
// ═══════════════════════════════════════════════════════════════════════════════

// ListDexes returns all available DEX names.
func (c *HyperCoreClient) ListDexes() ([]string, error) {
	result, err := c.call(context.Background(), "hl_listDexes", []any{})
	if err != nil {
		return nil, err
	}
	if arr, ok := result.([]any); ok {
		var dexes []string
		for _, d := range arr {
			if s, ok := d.(string); ok {
				dexes = append(dexes, s)
			}
		}
		return dexes, nil
	}
	return nil, nil
}

// ListMarkets returns all available markets, optionally filtered by DEX.
func (c *HyperCoreClient) ListMarkets(dex string) ([]any, error) {
	params := map[string]any{}
	if dex != "" {
		params["dex"] = dex
	}
	result, err := c.call(context.Background(), "hl_listMarkets", params)
	if err != nil {
		return nil, err
	}
	if arr, ok := result.([]any); ok {
		return arr, nil
	}
	return nil, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// ORDER QUERIES (User-specific)
// ═══════════════════════════════════════════════════════════════════════════════

// OpenOrders returns open orders for a user.
func (c *HyperCoreClient) OpenOrders(user string) ([]any, error) {
	result, err := c.call(context.Background(), "hl_openOrders", map[string]any{"user": user})
	if err != nil {
		return nil, err
	}
	if arr, ok := result.([]any); ok {
		return arr, nil
	}
	return nil, nil
}

// OrderStatus returns status of a specific order.
func (c *HyperCoreClient) OrderStatus(user string, oid int64) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_orderStatus", map[string]any{"user": user, "oid": oid})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// Preflight validates an order before signing.
func (c *HyperCoreClient) Preflight(coin string, isBuy bool, limitPx, sz, user string, reduceOnly bool, orderType map[string]any) (map[string]any, error) {
	params := map[string]any{
		"coin":       coin,
		"isBuy":      isBuy,
		"limitPx":    limitPx,
		"sz":         sz,
		"user":       user,
		"reduceOnly": reduceOnly,
	}
	if orderType != nil {
		params["orderType"] = orderType
	}
	result, err := c.call(context.Background(), "hl_preflight", params)
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// BUILDER FEE
// ═══════════════════════════════════════════════════════════════════════════════

// GetMaxBuilderFee returns maximum builder fee for a user-builder pair.
func (c *HyperCoreClient) GetMaxBuilderFee(user, builder string) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_getMaxBuilderFee", map[string]any{"user": user, "builder": builder})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// ORDER BUILDING (Returns unsigned actions for signing)
// ═══════════════════════════════════════════════════════════════════════════════

// BuildOrder builds an order action for signing.
func (c *HyperCoreClient) BuildOrder(coin string, isBuy bool, limitPx, sz, user string, reduceOnly bool, orderType map[string]any, cloid string) (map[string]any, error) {
	params := map[string]any{
		"coin":       coin,
		"isBuy":      isBuy,
		"limitPx":    limitPx,
		"sz":         sz,
		"user":       user,
		"reduceOnly": reduceOnly,
	}
	if orderType != nil {
		params["orderType"] = orderType
	}
	if cloid != "" {
		params["cloid"] = cloid
	}
	result, err := c.call(context.Background(), "hl_buildOrder", params)
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// BuildCancel builds a cancel action for signing.
func (c *HyperCoreClient) BuildCancel(coin string, oid int64, user string) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_buildCancel", map[string]any{"coin": coin, "oid": oid, "user": user})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// BuildModify builds a modify action for signing.
func (c *HyperCoreClient) BuildModify(coin string, oid int64, user string, limitPx, sz *string, isBuy *bool) (map[string]any, error) {
	params := map[string]any{"coin": coin, "oid": oid, "user": user}
	if limitPx != nil {
		params["limitPx"] = *limitPx
	}
	if sz != nil {
		params["sz"] = *sz
	}
	if isBuy != nil {
		params["isBuy"] = *isBuy
	}
	result, err := c.call(context.Background(), "hl_buildModify", params)
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// BuildApproveBuilderFee builds a builder fee approval for signing.
func (c *HyperCoreClient) BuildApproveBuilderFee(user, builder, maxFeeRate string, nonce int64) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_buildApproveBuilderFee", map[string]any{
		"user":       user,
		"builder":    builder,
		"maxFeeRate": maxFeeRate,
		"nonce":      nonce,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// BuildRevokeBuilderFee builds a builder fee revocation for signing.
func (c *HyperCoreClient) BuildRevokeBuilderFee(user, builder string, nonce int64) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_buildRevokeBuilderFee", map[string]any{
		"user":    user,
		"builder": builder,
		"nonce":   nonce,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// SENDING SIGNED ACTIONS
// ═══════════════════════════════════════════════════════════════════════════════

// SendOrder sends a signed order.
func (c *HyperCoreClient) SendOrder(action map[string]any, signature string, nonce int64) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_sendOrder", map[string]any{
		"action":    action,
		"signature": signature,
		"nonce":     nonce,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// SendCancel sends a signed cancel.
func (c *HyperCoreClient) SendCancel(action map[string]any, signature string, nonce int64) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_sendCancel", map[string]any{
		"action":    action,
		"signature": signature,
		"nonce":     nonce,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// SendModify sends a signed modify.
func (c *HyperCoreClient) SendModify(action map[string]any, signature string, nonce int64) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_sendModify", map[string]any{
		"action":    action,
		"signature": signature,
		"nonce":     nonce,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// SendApproval sends a signed builder fee approval.
func (c *HyperCoreClient) SendApproval(action map[string]any, signature string) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_sendApproval", map[string]any{
		"action":    action,
		"signature": signature,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// SendRevocation sends a signed builder fee revocation.
func (c *HyperCoreClient) SendRevocation(action map[string]any, signature string) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_sendRevocation", map[string]any{
		"action":    action,
		"signature": signature,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// WEBSOCKET SUBSCRIPTIONS (via JSON-RPC)
// ═══════════════════════════════════════════════════════════════════════════════

// Subscribe subscribes to a WebSocket stream via JSON-RPC.
func (c *HyperCoreClient) Subscribe(subscription map[string]any) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_subscribe", map[string]any{"subscription": subscription})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

// Unsubscribe unsubscribes from a WebSocket stream.
func (c *HyperCoreClient) Unsubscribe(subscription map[string]any) (map[string]any, error) {
	result, err := c.call(context.Background(), "hl_unsubscribe", map[string]any{"subscription": subscription})
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}
